### PR TITLE
Add a flag after the lock (permit) is acquired.

### LINKED ISSUE #94: Implementation of Locks with Hazelcast
Hi, there,

I just noticed that the implementation of Locks with Hazelcast is via `ISemaphore` which can be potentially release a permit multiple times. Though it is developer's obligation to use the `Semaphore` correctly, from the `Lock`'s point of view, it should be able to release more than one time, e.g. in multiple exception handler. But when it tries to acquire the lock, multiple `release()` should not result that the `Lock` can be acquired multiple times.

This could be a drawback of using `ISemaphore` to implement the `Lock`, while I am thinking the `Ignite`'s implementation using a queue with a cap (max size) is more appropriate.

Any ideas?

### PR DESCRIPTION
**Pull request by meggarr**
Add a flag to protect the lock from being released multiple times. This fixes #94.

Signed-off-by: Richard MENG <mengangr@foxmail.com>

### CODE REVIEW

File: src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java [hunk 1]
```diff
@@ -506,15 +507,22 @@ public void compareAndSet(long expected, long value, Handler<AsyncResult<Boolean
...
     private HazelcastLock(ISemaphore semaphore) {
       this.semaphore = semaphore;
+      this.released = new AtomicBoolean(false);
     }
 
     @Override
     public void release() {
       vertx.executeBlocking(future -> {
-        semaphore.release();
+        synchronized (released) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java]: Synchronization can be avoided and we should not run a task on the worker pool if released has been called already.

This should do:

```java
      if (released.compareAndSet(false, true)) {
        vertx.executeBlocking(future -> {
          semaphore.release();
          future.complete();
        }, false, null);
      }
```

Comment (meggarr) [reply src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java]: Is there any chance the `executeBlocking` will fail? I was thinking that it should set the flag right after the `release()`.

Comment (meggarr) [reply src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java]: Do you think it still needs a update? I am thinking it’s already good. If you insist, I can do it by the end of this week. (As I am out for several days.)

Comment (tsegismont) [reply src/main/java/io/vertx/spi/cluster/hazelcast/HazelcastClusterManager.java]: Yes. No worries. I'll go ahead and push the fix.

**Comment by tsegismont** [comment]
Closed via 96c4b69

**Pull request closed without merging**